modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java

```diff
@@ -157,7 +157,7 @@ MatrixStatsResults getResults() {
 
     @Override
     public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
-        if (results != null && results.getFieldCounts().keySet().isEmpty() == false) {
+        if (results != null) {
```

Comment:
I agree, after looking again `results` is never null once the aggregations are reduced. It can though be `null` at the transport level when shard results are sent over the wire before being reduced on the coordinating node.

It explains why `results` is randomly set to null in `InternalMatrixStats` because serialization is tested there. I reverted this change and only test the normal case.